Add migration guide

devendran-m opened the pull request:
This PR copies Michael's Migration Guide draft(GitBook version) to docs-redux. 

This PR also includes the pointers Ed highlighted in a couple of L1 and Sprint planning meetings. These sections are highlighted with the comment **"Placeholder—Core Team input required"**, which seeks feedback and input to expand the content needed for exchanges. 

**Note:** 
- The changes were tested locally, and the result is included in the PR.

**Screenshot:**
![image](https://github.com/user-attachments/assets/089a8e2e-3118-4cfd-aec4-e3a5c856266d)

**Test result PDF:**
[Casper_2.0_Migration_Guide_for_Crypto_Exchanges _ CasperDocs-Redux.pdf](https://github.com/user-attachments/files/18773263/Casper_2.0_Migration_Guide_for_Crypto_Exchanges._.CasperDocs-Redux.pdf)

davidatwhiletrue commented on condor/migration-guide.md:
```diff
@@ -0,0 +1,112 @@
+| |The appropriate SDK to upgrade to is the lates v5.x release found [here](https://github.com/casper-ecosystem/casper-js-sdk/releases). |
+| | Common use case examples (connect to RPC, send a transfer, etc.) can be found in [Casper JS SDK 5.0 README](https://github.com/ casper-ecosystem/casper-js-sdk/tree/feat-5.0.0?tab=readme-ov-file#usage-examples) |
+| |To upgrade your integration, follow the detailed [V2 to V5 Migration Guide](https://github.com/casper-ecosystem/casper-js-sdk/blob/feat-5.0.0/resources/migration-guide-v2-v5.md) | 
+| **#C .NET** / **Go** / **Java** | While it may not be strictly necessary to upgrade your integration if all you do is submit `Deploys` to the network (e.g. transfers in and out of accounts), it is STRONGLY RECOMMENDED to upgrade to the new SDK now, to experience a seamless upgrade to Casper 2.0 and beyond.|
```
@devendran-m 
Content for C# SDK:
- The appropriate SDK to upgrade to is the latest v3.x release found [[here](https://github.com/make-software/casper-net-sdk/releases)](https://github.com/make-software/casper-net-sdk/releases).
- Typical use case examples (connect to RPC, send a transfer, etc.) can be found in [[the Examples folder](https://github.com/make-software/casper-net-sdk/tree/casper-2.0.0/Docs/Examples)](https://github.com/make-software/casper-net-sdk/tree/casper-2.0.0/Docs/Examples).
- To upgrade your integration, follow the document [[Migration from Casper .NET SDK v2.x to v3.0](https://github.com/make-software/casper-net-sdk/blob/casper-2.0.0/Docs/Articles/Casper20MigrationGuide.md)](https://github.com/make-software/casper-net-sdk/blob/casper-2.0.0/Docs/Articles/Casper20MigrationGuide.md)

devendran-m replied:
Thanks @davidatwhiletrue , 

I've incorporated your comments in the commit - [ff82e88](https://github.com/casper-network/docs-redux/pull/5/commits/ff82e885077f753148d455964421e05cf0044494)

Please review and let me know in case of any other changes. 

cc: @KMCreatesWorlds

davidatwhiletrue commented on condor/migration-guide.md:
```diff
@@ -0,0 +1,112 @@
+| | Common use case examples (connect to RPC, send a transfer, etc.) can be found in [Casper JS SDK 5.0 README](https://github.com/ casper-ecosystem/casper-js-sdk/tree/feat-5.0.0?tab=readme-ov-file#usage-examples) |
+| |To upgrade your integration, follow the detailed [V2 to V5 Migration Guide](https://github.com/casper-ecosystem/casper-js-sdk/blob/feat-5.0.0/resources/migration-guide-v2-v5.md) | 
+| **#C .NET** / **Go** / **Java** | While it may not be strictly necessary to upgrade your integration if all you do is submit `Deploys` to the network (e.g. transfers in and out of accounts), it is STRONGLY RECOMMENDED to upgrade to the new SDK now, to experience a seamless upgrade to Casper 2.0 and beyond.|
+| **Other** | If you are using an SDK not listed in this chapter, please contact the Casper Exchange Support Team.|
```
@devendran-m 
Content for Go SDK:

- The appropriate SDK to upgrade to is the latest v2.x release found [[here](https://github.com/make-software/casper-go-sdk/releases)](https://github.com/make-software/casper-go-sdk/releases).
- Typical use case examples (connect to RPC, send a transfer, etc.) can be found in the [[README](https://github.com/make-software/casper-go-sdk/blob/master/README.md)](https://github.com/make-software/casper-go-sdk/blob/master/README.md).

devendran-m replied:
Thanks @davidatwhiletrue , 

I've incorporated your comments in the commit - [ff82e88](https://github.com/casper-network/docs-redux/pull/5/commits/ff82e885077f753148d455964421e05cf0044494)

Please review and let me know in case of any other changes. 

cc: @KMCreatesWorlds

davidatwhiletrue reviewed (commented):
@devendran-m I've added to comments content for .NET SDK and Go SDK .

devendran-m commented:
Hi @ihor , 
This is the migration guide PR.

darthsiroftardis reviewed (commented):
So overall this document feels incomplete to me. 

1. We need to mention that 1.x structures like Deploys are being deprecated, so while they will continue to work 2.x, they're due to be phased out and exchanges should understand that new types like `TransactionV1` are the go forward.
2. Additionally, the sidecar needs mentioning as well, especially if certain exchanges run their own node, which means we need to guide them on the fact that RPC component now resides on the sidecar, which this document doesn't mention
3. The document keeps mentioning that its 'recommended' to be using the latest SDK release. It should be clear that updating to a new SDK will become necessary when addressable entity becomes enabled, which will require the exchanges to be using an SDK that is compatible with the Entity go forward.

zajko commented on condor/migration-guide.md:
```diff
@@ -0,0 +1,119 @@
+
+The Casper Network is in the process of upgrading to Casper 2.0, which introduces breaking changes to the node software, the protocol and the interfaces associated with interacting with the network. In order to maintain compatibility with Casper Network after the upgrade of mainnet, tentatively scheduled for March 2025, crypto exchanges that offer $CSPR pairs, allow for $CSPR deposits and withdra ...
+
+More information about the specific changes in, and features introduced by Casper 2.0 can be found here (LINK).
```
Should the link point to the changelog? when will we instantiate it?

ihor replied:
Typo ` in,` should be removed together with the comma

zajko commented on condor/migration-guide.md:
```diff
@@ -0,0 +1,119 @@
+
+### RPC Integration
+Casper 2.0 implements a number of changes to the JSON RPC interface of the node, which includes new methods, renamed methods and methods with a change in response format. 
+Importantly, the block structure is different in Casper 2.0 vs. Casper 1.x, and `Deploys` have been renamed to `Transactions`. 
```
"`Deploys` have been renamed to `Transactions`" is misleading. `Transactions` are a higher-order concept which encapsulate both `Deploys` and `V1` transactions. `Transaction` is a polymorphic data structure that, for now is either a `Deploy` (which should be compliant with the 1.x definition) or a `V1`

zajko replied:
Also,
"Casper 2.0 implements a number of changes to the JSON RPC interface of the node"
The node no longer has a json rpc interface.

devendran-m replied:
Thanks, @zajko , 
I've incorporated the possible and required changes in the latest commit `3f8c52248dc059c617f58ffd0b6aa6b1331c9a04`.

zajko commented on condor/migration-guide.md:
```diff
@@ -0,0 +1,119 @@
+Casper 2.0 implements a number of changes to the JSON RPC interface of the node, which includes new methods, renamed methods and methods with a change in response format. 
+Importantly, the block structure is different in Casper 2.0 vs. Casper 1.x, and `Deploys` have been renamed to `Transactions`. 
+
+At the same time, legacy (Casper 1.x) deploys will continue to be accepted by Casper 2.0, hence if your entire integration comprises only the submission of Deploys (such as transfers in and out of accounts), you may not have to make any changes prior to mainnet activation of Casper 2.0. 
```
"you may not have to make any changes prior to mainnet activation of Casper 2.0. "
That statement is not precise. Although the `account-put-deploy` JSON RPC method should accept the same structure of the input data, the result of "info_get_deploy" is not backwards compatible with 1.x.
So the only way an exchange would not need to do any changes is if they wouldn't care for the results of an execution (?)

devendran-m replied:
Thanks, @zajko , 
I've incorporated the possible and required changes in the latest commit `3f8c52248dc059c617f58ffd0b6aa6b1331c9a04`.

zajko commented on condor/migration-guide.md:
```diff
@@ -0,0 +1,119 @@
+
+At the same time, legacy (Casper 1.x) deploys will continue to be accepted by Casper 2.0, hence if your entire integration comprises only the submission of Deploys (such as transfers in and out of accounts), you may not have to make any changes prior to mainnet activation of Casper 2.0. 
+
+➡️ To assess your next steps, please evaluate your integration agains the 1.x -> 2.0 JSON RPC changes detailed in this document. 
```
Is the json rpc changes section TBD or are they somewhere else?

devendran-m replied:
Thanks, @zajko , 
I've incorporated the possible and required changes in the latest commit `3f8c52248dc059c617f58ffd0b6aa6b1331c9a04`. I have linked the existing Casper 2.0 page that explains the 1.x vs 2.0 JSON RPC changes. Please let me know if that's the appropriate link we can use in this context.

devendran-m commented:
> I agree with what @darthsiroftardis wrote, additionally:
> 
> * we need to explain that an operator which wants to have a forward facing API needs to turn on binary-port (since sidecar uses it). But in the same time the binary-port doesn't need to be a public port (it only needs to be exposed to sidecar)
> * I'm not sure what are we aiming at in this document. If we want to provide a minimal guide on how to migrate - I would focus on explaining how to use 1.x deploys in a 2.x node - stressing that `sidecars` `account_put_deploy` should be compatible with 1.x node `account_put_deploy`, but sidecars `info_get_deploy` isn't 100% compatible with 1.x nodes `info_get_deploy`
> * We need a section on the SSE feed changes


Thanks, @zajko , 
I've incorporated the possible and required changes in the latest commit `3f8c52248dc059c617f58ffd0b6aa6b1331c9a04`.

ihor commented on condor/migration-guide.md:
```diff
@@ -0,0 +1,193 @@
+# Casper 2.0 Migration Guide for Crypto Exchanges
+
+The Casper Network is in the process of upgrading to Casper 2.0, which introduces breaking changes to the node software, the protocol and the interfaces associated with interacting with the network. In order to maintain compatibility with Casper Network after the upgrade of mainnet, tentatively scheduled for March 2025, crypto exchanges that offer $CSPR pairs, allow for $CSPR deposits and withdra ...
```
Let's shorten `the interfaces associated with interacting with the network` to `the API interfaces`

ihor replied:
Technically, there should be an Oxford comma after `the protocol` in `the node software, the protocol, and the interfaces`, but I don't want to start a fight here. Let's just make sure we use the oxford comma consistently.

ihor commented on condor/migration-guide.md:
```diff
@@ -0,0 +1,193 @@
+# Casper 2.0 Migration Guide for Crypto Exchanges
+
+The Casper Network is in the process of upgrading to Casper 2.0, which introduces breaking changes to the node software, the protocol and the interfaces associated with interacting with the network. In order to maintain compatibility with Casper Network after the upgrade of mainnet, tentatively scheduled for March 2025, crypto exchanges that offer $CSPR pairs, allow for $CSPR deposits and withdra ...
```
Inconsistent capitalization of `mainnet`. It's capitalized in all (most of?) other places in the documentation `Mainnet`

ihor replied:
Can we shorten `crypto exchanges that offer $CSPR pairs, allow for $CSPR deposits and withdrawals and/or allow their users to stake $CSPR` to `crypto exchanges that work with the $CSPR token`?

ihor replied:
> This migration guide will help you determine whether, and if so - how, to migrate your integration in prepartion for the network upgrade.

1. Typo `preparation`
2. Typo hyphen `-` instead of dash `–`
3. Let's replace `migration` here with `update`

ihor commented on condor/migration-guide.md:
```diff
@@ -0,0 +1,193 @@
+
+More information about the specific changes in, and features introduced by Casper 2.0 can be found here (LINK).
+
+This migration guide aims at listing down the key information required for exchanges to make the necessary changes on their side to become Casper 2.0 ready.
```
We can remove this paragraph w/o loosing information or clarity.

ihor commented on condor/migration-guide.md:
```diff
@@ -0,0 +1,193 @@
+| **Indirect** | Some exchanges do not integrate with Casper Network directly but leverage liquidity provided by other exchanges through their APIs | No migration required (assuming the downstream exchange migrates on time, if needed) |
+| **[Run own node](#)** | Some exchanges run their own node and interface with its RPC methods | Exchange needs to upgrade the node software, and participate in the network upgrade process. Additionally, exchange needs to ensure the RPC methods they rely on are either still supported in 2.0 or migrate to new RPC methods |
+| **[RPC integration](#)** | Some exchanges use public node infrastructure and interface with those via RPC | Exchange needs to ensure the RPC methods they rely on are either still supported in 2.0, or migrate to new RPC methods |
+| **[SDK integration](#)** | Some exchanges use the publicly available Casper SDKs (e.g. JavaScript SDK, C#.NET SDK, Go SDK, etc.) | Depending on the SDK language (e.g. JS, Go) and the methods you use, you may or may not need to migrate prior to mainnet launch |
```
"This migration guide will help you determine whether, and if so - how, to migrate your integration in prepartion for the network upgrade.

More information about the specific changes in, and features introduced by Casper 2.0 can be found here (LINK).

This migration guide aims at listing down the key information required for exchanges to make the necessary changes on their side to become Casper 2.0 ready.

Should I migrate my integration?

Exchanges have different levels of integration with Casper Network, and offer different subsets of functionality relating to $CSPR to their user base. Depending on the specifics of your integration, you may or may not have to migrate your integration prior to the mainnet launch of Casper 2.0. The table below helps you determine your next steps, if any."

The above piece of text contains repeated information. Similarly, the table seems to be unnecessarily verbose. What about replacing them with the following?

This guide should help you determine if any changes are needed on your end and, if so, how to update your integration in preparation for the network upgrade.

**Should I update my integration?**
| Integration type | Required actions|
|---|---|
| Indirect integration through a 3rd party partner | No migration required, assuming the partner updates their integration |
| Running a node | Node software upgrade is required |
| Direct RPC API interaction | Adjustments to new response structures is required. Deprecated methods should be replaced with their newer alternatives |
| Interaction through an SDK | Upgrade to a Casper 2.0 compatible version is required |

More information about the specific changes in, and features introduced by Casper 2.0 can be found [here](TODO).

ihor commented on condor/migration-guide.md:
```diff
@@ -0,0 +1,193 @@
+
+**SSE (Server Side Events) Server**
+
+Casper 2.0 introduces a realtime API of events, emitted by the node when processing blocks and/or transactions.
```
As mentioned in my previous comment, the section above doesn't belong here.

ihor commented on condor/migration-guide.md:
```diff
@@ -0,0 +1,193 @@
+
+➡️ To assess your next steps, please evaluate your integration against the 1.x -> 2.0 JSON RPC changes detailed in the link [here](https://docs.casper.network/condor/jsonrpc#json-rpc-schema-definitions). 
+
+Please refer to this [page](https://docs.casper.network/condor/jsonrpc) for more details on **casper-sidecar** and **JSON RPC** changes.
```
The RPC integration section doesn't provide much information about what has to be changed, essentially sending the reader to check the RPC schema by themselves. 

The above has to be rewritten to include the list of changes in the RPC protocol divided into two parts:
1. Change in types with correspondences between the semantically identical types (`Deploy` and `TransactionV1`, `Block` and `BlockV2`)
2. Changes in the RPC methods that clearly describe which methods are deprecated and what are their newer alternatives.

ihor replied:
Additionally, this section is a very heavy lift. Do we know of case where somebody integrates with the RPC interface directly w/o an SDK?

ihor commented on condor/migration-guide.md:
```diff
@@ -0,0 +1,193 @@
+### SDK Integration
+Depending on which SDK language (e.g. JavaScript, Go) you use, and which methods you interface with, you might not have to immediately migrate your integration. However, in order to be safe and future-proof, it is strongly recommended to upgrade your SDK to the latest version applicable to your stack. For most languages, this can be done immediately, as the SDKs maintain both forward (2.x) and ba ...
+
+While it may not be strictly necessary to upgrade your integration if all you do is submit Deploys to the network (e.g. transfers in and out of accounts), it is STRONGLY RECOMMENDED to upgrade to the new SDK now, to experience a seamless upgrade to Casper 2.0 and beyond.
```
The two paragraphs above kind of repeat themselves that "you may or may not need to upgrade". We should replace the intro with more specifics of how new SDKs operate to help the reader understand the changes and make the decision. We could replace it with something like this:

Casper 2.0 introduces changes essential for the network’s future evolution, which also include modifications to data structures that impact API responses. Since historical blockchain data cannot be altered due to cryptographic signatures, both legacy and new data structures will coexist in Casper 2.0 and be returned by the API.

To simplify the transition for developers, we prepared new backward and forward-compatible Casper SDKs that provide a uniform data interface across both Casper 1.5.x and Casper 2.0. This ensures that developers can work with consistent data types, regardless of the network version in which the data was created. 

Upgrading to a new version of Casper SDKs makes it possible to have one integration that guarantees future compatibility with Casper 2.0 while maintaining the existing compatibility with Casper 1.5. The network upgrade will require no changes in the consumer code, ensuring no disruption in the service.

ihor commented on condor/migration-guide.md:
```diff
@@ -0,0 +1,193 @@
+> 
+> Updating to a Casper 2.0-compatible SDK will be necessary when the **addressable entity** is enabled.
+> 
+> Once the **addressable entity** is enabled, exchanges will be required to use an SDK compatible with the Entity.
```
We should not include not-enabled functionality in the migration guide. It unnecessarily overloads the reader w/o providing any clear value.

ihor commented on condor/migration-guide.md:
```diff
@@ -0,0 +1,193 @@
+| |Typical use case examples (connect to RPC, send a transfer, etc.) can be found in the [README](https://github.com/make-software/casper-go-sdk/blob/master/README.md). | 
+| **Java** | The appropriate SDK to upgrade to is the latest v2.x release can be found [here](https://github.com/casper-network/casper-java-sdk/releases).|
+| |Casper Java SDK in Maven Repository can be found in the link [here](https://central.sonatype.com/artifact/network.casper/casper-java-sdk/2.7.0-BETA.4/overview). | 
+| **Other** | If you are using an SDK not listed in this chapter, please contact the Casper Exchange Support Team.|
```
Is there any reason why we use a table here? It's not formatted conventionally. Having subsections for each language would probably be more readable:

<img width="983" alt="Screenshot 2025-03-06 at 13 53 08" src="https://github.com/user-attachments/assets/b0746b69-2085-4a94-b8c9-d990307cfe9a" />

Additionally, the last row says to contact the Casper team but doesn't provide any way how to do that.

devendran-m commented:
Hi @ihor , 
cc: @zajko , @darthsiroftardis , @EdHastingsCasperAssociation 

I've addressed most of your comments in the latest commit - 1b2d0eca76b4098ca7fdc3b3c1e9800e8ccb94af. 
Regarding the information about the sidecar, transactions, and the difference between 1.x and 2.0, I have added the details shared by Jakub and Karan in their earlier review cycle. I also request that the core team review and share their input/feedback to elaborate the guide.

zajko commented on condor/migration-guide.md:
```diff
@@ -1,56 +1,135 @@
+The Casper Network is in the process of upgrading to Casper 2.0, which introduces breaking changes to the node software, the protocol, and the API Interfaces. 
 
-More information about the specific changes in, and features introduced by Casper 2.0 can be found here (LINK).
+In order to maintain compatibility with Casper Network after the upgrade of Mainnet, which is tentatively scheduled for end of March/mid-April 2025, crypto exchanges that offer $CSPR pairs, allow for $CSPR deposits and withdrawals and/or allow their users to stake $CSPR, may need to upgrade their integration with Casper Network. This migration guide will help you determine whether, and if so - ho ...
```
We're aiming at a mainnet upgrade in April?

zajko commented on condor/migration-guide.md:
```diff
@@ -0,0 +1,193 @@
+If you run your own node, your migration experience will likely involve two distinct paths. 
+
+On the one hand, as a node operator you will participate in the actual mainnet upgrade of Casper Network, which will take place a specific date and time (you will have a number of days to stage the upgrade in advance of the activation point), while on the other hand you will have to possibly migrate the method by which you interface with your own node. Continue with the migration details provided ...
+
```
"the required libraries exist only for the Rust language."
If you're referring to `casper-binary-port-client` - we don't intend to release it in the near future  AFAIK so I wouldn't mention it.

I would also omit details like storing SSE events since it's not relevant to the migration of exchanges.

mssteuer commented:
I'm merging this for now. Further changes can be performed in another PR
